timm/models/pe.py

```diff
@@ -0,0 +1,693 @@
+from collections import OrderedDict
+from functools import partial
+from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple, Union, Literal
+
+import torch
+import torch.nn as nn
+from torch import nn, Tensor, broadcast_tensors, einsum
+from torch.nn import functional as F
+from torch.nn import Module, ModuleList
+from torch.nn.init import constant_, xavier_normal_, xavier_uniform_
+from torch.nn.parameter import Parameter
+from torch.amp import autocast
+from torch.utils.checkpoint import checkpoint
+from torch import pi
+
+### Import timm layers
+from timm.layers import (
+    PatchEmbed,
+    Mlp,
+    DropPath,
+    AttentionPoolLatent,
+    RmsNorm,
+    PatchDropout,
+    SwiGLUPacked,
+    SwiGLU,
+    trunc_normal_,
+    lecun_normal_,
+    resample_patch_embed,
+    resample_abs_pos_embed,
+    use_fused_attn,
+    get_act_layer,
+    get_norm_layer,
+    LayerType,
+    LayerScale,
+)
+
+# from timm.layers import RotaryEmbeddingCat, RotaryEmbedding # not compatible
+from timm.data import IMAGENET_INCEPTION_MEAN, IMAGENET_INCEPTION_STD
+
+from ._builder import build_model_with_cfg
+from ._features import feature_take_indices
+from ._registry import generate_default_cfgs, register_model, register_model_deprecations
+
+
+__all__ = ['PE']
+
+
+######## PE Rope (Simplified) ########
+class RotaryEmbedding(Module):
+    def __init__(
+        self,
+        dim,
+        freqs_for: Union[Literal["lang"], Literal["pixel"], Literal["constant"]] = "lang",        
+        theta=10000,
+        max_freq=10,
+        num_freqs=1,
+        learned_freq=False,              
+        theta_rescale_factor=1.0,
+    ):
+        super().__init__()
+        # proposed by reddit user bloc97, to rescale rotary embeddings to longer sequence length without fine-tuning
+        # has some connection to NTK literature
+        # https://www.reddit.com/r/LocalLLaMA/comments/14lz7j5/ntkaware_scaled_rope_allows_llama_models_to_have/
+        theta *= theta_rescale_factor ** (dim / (dim - 2))
+        if freqs_for == "lang":
+            freqs = 1.0 / (theta ** (torch.arange(0, dim, 2)[: (dim // 2)].float() / dim))
+        elif freqs_for == "pixel":
+            freqs = torch.linspace(1.0, max_freq / 2, dim // 2) * pi
+        elif freqs_for == "constant":
+            freqs = torch.ones(num_freqs).float()
+        self.freqs = nn.Parameter(freqs, requires_grad=learned_freq)
+
+    def forward(self, t: Tensor): #, seq_len=None, offset=0):
+        freqs = self.freqs
+        freqs = t.type(freqs.dtype).unsqueeze(-1) * freqs
+        freqs = freqs.repeat_interleave(2, dim=-1)
+        return freqs
+
+
+class Rope2D(Module):
+    def __init__(self, dim, grid_size, use_cls_token=False):
```

Comment:
This module should be marked non traceable to pass FX tests as it looks like the `if t.ndim == 3` will break tracing

See eg https://github.com/huggingface/pytorch-image-models/blob/c8c4f256b8c279575ed87012f38620a4cd25df57/timm/models/xcit.py#L33 for use of notrace decorator